# Lista 4

In [38]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.neighbors import KNeighborsRegressor
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import LogisticRegression

## Zadanie 1

In [123]:
data = pd.read_csv('advertising.csv')

In [124]:
data

TV  Radio  Newspaper  Sales
0    230.1   37.8       69.2   22.1
1     44.5   39.3       45.1   10.4
2     17.2   45.9       69.3   12.0
3    151.5   41.3       58.5   16.5
4    180.8   10.8       58.4   17.9
..     ...    ...        ...    ...
195   38.2    3.7       13.8    7.6
196   94.2    4.9        8.1   14.0
197  177.0    9.3        6.4   14.8
198  283.6   42.0       66.2   25.5
199  232.1    8.6        8.7   18.4

[200 rows x 4 columns]

Uzywamy TV, Radio i sales

In [125]:
X = data[['TV', 'Radio']].values.reshape(-1, 2)
y = data['Sales'].values.reshape(-1)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

scaler = StandardScaler()
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [126]:
res = {'k': [],
       'mse': [],
       'mse_scaled': []
       }
for k in range(1, 16, 2):
    model = KNeighborsRegressor(n_neighbors=k)
    model.fit(X_train, y_train)
    mse = mean_squared_error(y_test, model.predict(X_test))

    model_scaled = KNeighborsRegressor(n_neighbors=k)
    model_scaled.fit(X_train_scaled, y_train)
    mse_scaled = mean_squared_error(y_test, model_scaled.predict(X_test_scaled))
    res['k'].append(k)
    res['mse'].append(mse)
    res['mse_scaled'].append(mse_scaled)

In [127]:
pd.DataFrame.from_dict(res)

k       mse  mse_scaled
0   1  2.855500    2.659750
1   3  1.408722    1.554250
2   5  1.386360    1.324080
3   7  1.688974    1.542750
4   9  1.830398    1.451515
5  11  2.028785    1.693446
6  13  2.314337    1.895891
7  15  2.423810    2.044570

## Wnioski

Regesja liniowa dala MSE 2.4 dla dwoch zmiennych

Mse bez skalowania wypada minimalnie gorzej od tego ze skalowaniem. Dla kilku zmiennych o roznych rozkladach zawsze warto zeskalowac wartosci. KNN poradzil sobie w tym zadaniu lepiej od regresji liniowej (MSE 1.32 vs 2.4).


## Zadanie 2

In [121]:
data = pd.read_csv('bank.csv')

In [122]:
data

age          job  marital  education default  balance housing loan  \
0       59       admin.  married  secondary      no     2343     yes   no   
1       56       admin.  married  secondary      no       45      no   no   
2       41   technician  married  secondary      no     1270     yes   no   
3       55     services  married  secondary      no     2476     yes   no   
4       54       admin.  married   tertiary      no      184      no   no   
...    ...          ...      ...        ...     ...      ...     ...  ...   
11157   33  blue-collar   single    primary      no        1     yes   no   
11158   39     services  married  secondary      no      733      no   no   
11159   32   technician   single  secondary      no       29      no   no   
11160   43   technician  married  secondary      no        0      no  yes   
11161   34   technician  married  secondary      no        0      no   no   

        contact  day month  duration  campaign  pdays  previous poutcome  \
0       unknown    5   may      1042         1     -1         0  unknown   
1       unknown    5   may      1467         1     -1         0  unknown   
2       unknown    5   may      1389         1     -1         0  unknown   
3       unknown    5   may       579         1     -1         0  unknown   
4       unknown    5   may       673         2     -1         0  unknown   
...         ...  ...   ...       ...       ...    ...       ...      ...   
11157  cellular   20   apr       257         1     -1         0  unknown   
11158   unknown   16   jun        83         4     -1         0  unknown   
11159  cellular   19   aug       156         2     -1         0  unknown   
11160  cellular    8   may         9         2    172         5  failure   
11161  cellular    9   jul       628         1     -1         0  unknown   

      deposit  
0         yes  
1         yes  
2         yes  
3         yes  
4         yes  
...       ...  
11157      no  
11158      no  
11159      no  
11160      no  
11161      no  

[11162 rows x 17 columns]

In [ ]:
data['job']